### PR TITLE
cephadm: deal with ambiguity within normalize_image_digest

### PR DESCRIPTION
**Pull request by sebastian-philipp**
Fixes: https://tracker.ceph.com/issues/53594

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/tox.ini [hunk 1]
```diff
@@ -161,6 +162,7 @@ modules =
 commands =
     flake8 --config=tox.ini {posargs} \
       {posargs:{[testenv:flake8]modules}}
+    bash -c 'test $(git ls-files cephadm | grep ".py$" | grep -v tests | xargs grep "docker.io" | wc -l) == 12'
```
Comment (adk3798) [inline src/pybind/mgr/tox.ini:165]: what's the reason behind testing how many times docker.io appears in the codebase (and why are we assuming it should stay at this particular number, https://github.com/ceph/ceph/pull/43901 for example would change it)?

File: src/pybind/mgr/cephadm/upgrade.py [hunk 2]
```diff
@@ -22,16 +22,29 @@
...
+    >>> normalize_image_digest('localhost/ceph', 'docker.io')
+    'localhost/ceph'
+    """
+    known_shortnames = [
+        'ceph/ceph',
+        'ceph/daemon',
+        'ceph/daemon-base',
+    ]
+    for image in known_shortnames:
+        if digest.startswith(image):
```
Comment (adk3798) [inline src/pybind/mgr/cephadm/upgrade.py:46]:
```suggestion
   if any(digest.startswith(image) for image in known_shortnames):
```

NIT, saves a line if you like it, if not current form works fine. Same for the one in the binary.

File: src/pybind/mgr/cephadm/module.py [hunk 3]
```diff
@@ -282,7 +282,7 @@ class CephadmOrchestrator(orchestrator.Orchestrator, MgrModule,
             'registry_url',
             type='str',
             default=None,
-            desc='Custom repository url'
+            desc='Registry url for login purposes. This is not the default registry'
```
Comment (adk3798) [inline src/pybind/mgr/cephadm/module.py:285]: Can we also change the "registry_username" and "registry_password" descs to match. I.e. "Registry username for login purposes . . ."

**Review by adk3798 — APPROVED** [review]
LGTM, some optional nitpicks if you have time

File: src/pybind/mgr/tox.ini (hunk 1 above)
Comment (sebastian-philipp) [reply src/pybind/mgr/tox.ini:165]: I'd like to allow folks to change this number in their downstreams in order to make sure no new references to docker.io are creeping in unnoticed.

**Comment by sebastian-philipp** [comment]
```
                        'to_image', None)) == 'Initiating upgrade to docker.io/to_image'
E                   AssertionError: assert 'Initiating u...e to to_image' == 'Initiating u...r.io/to_image'
E                     - Initiating upgrade to docker.io/to_image
E                     ?                       ----------
E                     + Initiating upgrade to to_image

cephadm/tests/test_upgrade.py:59: AssertionError
```

File: src/pybind/mgr/tox.ini (hunk 1 above)
Comment (adk3798) [reply src/pybind/mgr/tox.ini:165]: makes sense

**Review by liewegas — APPROVED** [review]
lgtm

**Comment by sebastian-philipp** [comment]
https://pulpito.ceph.com/swagner-2022-01-06_10:04:35-orch:cephadm-wip-swagner-testing-2022-01-05-1435-distro-default-smithi/

somehow one of the PRs in this run broke ingress. Maybe this one?

**Comment by adk3798** [comment]
https://pulpito.ceph.com/melissali-2022-01-12_17:49:26-rados:cephadm-wip-swagner-testing-2022-01-11-1258-distro-basic-smithi/

Unrelated failures:

-----

`Command failed on smithi089 with status 1: 'CEPH_REF=master CEPH_ID="0" PATH=$PATH:/usr/sbin adjust-ulimits ceph-coverage /home/ubuntu/cephtest/archive/coverage /home/ubuntu/cephtest/virtualenv/bin/cram -v -- /home/ubuntu/cephtest/archive/cram.client.0/*.t'`

Not aware of a tracker for this but it has also happened on other runs as well and only on centos8.stream so has no correlation with this change

-----

Multiple dead jobs testing ingress, unrelated to this change

**Pull request merged**